### ISSUE TITLE
Liquid haskell receives SIGPIPE, mathsat driver

### ISSUE THREAD

**Issue by formrre** [ucsd-progsys/liquidhaskell#1020]
Hi,
while working with liquid haskell the process tends to receive SIGPIPE and die with "fd:11: hFlush: resource vanished"... fd:11: hFlush: resource vanished (Broken pipe)"
This happens ocassionaly without changing the file. My installation is built out of the develop branch, and the bug is reproducible on both Fedora and Debian, this happens when using the mathsat driver.
See the attached strace.
```
rt_sigprocmask(SIG_BLOCK, [HUP INT QUIT TERM WINCH], [], 8) = 0
timer_settime(0, 0, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=0, tv_nsec=0}}, NULL) = 0
vfork()                                 = 4000
close(10)                               = 0
fcntl(11, F_SETFD, FD_CLOEXEC)          = 0
close(13)                               = 0
fcntl(12, F_SETFD, FD_CLOEXEC)          = 0
close(15)                               = 0
fcntl(14, F_SETFD, FD_CLOEXEC)          = 0
close(17)                               = 0
fcntl(16, F_SETFD, FD_CLOEXEC)          = 0
read(16, "", 4)                         = 0
close(16)                               = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
timer_settime(0, 0, {it_interval={tv_sec=0, tv_nsec=10000000}, it_value={tv_sec=0, tv_nsec=10000000}}, NULL) = 0
fcntl(11, F_GETFL)                      = 0x1 (flags O_WRONLY)
fcntl(11, F_SETFL, O_WRONLY|O_NONBLOCK) = 0
ioctl(11, TCGETS, 0x7ffdd07cef70)       = -1 ENOTTY (Inappropriate ioctl for device)
fcntl(12, F_GETFL)                      = 0 (flags O_RDONLY)
fcntl(12, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
ioctl(12, TCGETS, 0x7ffdd07cef70)       = -1 ENOTTY (Inappropriate ioctl for device)
fcntl(14, F_GETFL)                      = 0 (flags O_RDONLY)
fcntl(14, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
ioctl(14, TCGETS, 0x7ffdd07cef70)       = -1 ENOTTY (Inappropriate ioctl for device)
mkdir("/home/user/liquidhaskell/newLiquid/.liquid", 0777) = -1 EEXIST (File exists)
stat("/home/user/liquidhaskell/newLiquid/.liquid", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
open("/home/user/liquidhaskell/newLiquid/.liquid/List.hs.smt2", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = 10
fstat(10, {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
ftruncate(10, 0)                        = 0
ioctl(10, TCGETS, 0x7ffdd07cef70)       = -1 ENOTTY (Inappropriate ioctl for device)
write(11, "(define-sort Elt () Int)\n", 25) = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=3995, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = -1 EPIPE (Broken pipe)
futex(0x5a29ebc, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x5a29eb8, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
futex(0x5a29ee8, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x4d69df8, FUTEX_WAKE_PRIVATE, 1) = 1
rt_sigprocmask(SIG_BLOCK, [WINCH], [], 8) = 0
rt_sigaction(SIGWINCH, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f3250122c30}, NULL, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
write(2, "liquid: ", 8liquid: )                 = 8
write(2, "fd:11: hFlush: resource vanished"..., 46fd:11: hFlush: resource vanished (Broken pipe)) = 46
```

**Comment by ranjitjhala** [comment]
Oops, sorry about that!

Usually this happens when the SMT solver has crashed for some reason.

1. Is there a code snippet you can send us so we can reproduce?

2. Does it happen with Z3 too?

**Comment by formrre** [comment]
Hi,
1. here is a minimal snippet; no CHECKBINDER annotations necessary. Also this one says "OOPS, hClose breaks: ctxCoutfd:11: hClose: resource vanished (Broken pipe)
liquid: fd:11: hFlush: resource vanished (Broken pipe)"

`module List(List,empty) where
data List a = Emp
            | (:+:) a (List a)
              deriving (Eq, Ord, Show)
{-@ measure size      :: List a -> Int
    size (Emp)        = 0
    size ((:+:) x xs) = 1 + size xs
  @-}

{-@ invariant {v:List a | 0 <= size v} @-}

{-@ type ListN a N  = {v:List a | size v = N} @-}

{-@ empty :: ListN a 0 @-}
empty = Emp`


{-@ invariant {v:List a | 0 <= size v} @-}

{-@ type ListN a N  = {v:List a | size v = N} @-}

{-@ empty :: ListN a 0 @-}
empty = Emp`
    size ((:+:) x xs) = 1 + size xs
  @-}

{-@ invariant {v:List a | 0 <= size v} @-}

{-@ type ListN a N  = {v:List a | size v = N} @-}

{-@ empty :: ListN a 0 @-}
empty = Emp`

2. I have encountered slightly different error with z3 - less reproducible, yet it is still quite annoying, I will try to reproduce it and fill in another bug.

**Comment by formrre** [comment]
Minimal snippet, also this one says:
"OOPS, hClose breaks: ctxCoutfd:11: hClose: resource vanished (Broken pipe)
 liquid: fd:11: hFlush: resource vanished (Broken pipe)
`module List(List,empty) where
data List a = Emp
            | (:+:) a (List a)
              deriving (Eq, Ord, Show)

{-@ measure size      :: List a -> Int
    size (Emp)        = 0
    size ((:+:) x xs) = 1 + size xs
  @-}

{-@ invariant {v:List a | 0 <= size v} @-}

{-@ type ListN a N  = {v:List a | size v = N} @-}

{-@ empty :: ListN a 0 @-}
empty = Emp`

There is slightly different issue with Z3 for example the above code does not finish solving and finishes after Uniqify & Rename with Working 0% and liquid: fd:12: hGetLine: end of file
Edit: I have pasted wrong code in here

**Comment by ranjitjhala** [comment]
Can you check that you have `mathsat` (and/or `z3` on your path?) When I run

```
$ stack exec -- liquid tests/todo/T1020.hs --smt=mathsat
```

I also get a 

```
OOPS, hClose breaks: ctxCoutfd:16: hClose: resource vanished (Broken pipe)
```

but its because I don't have `mathsat` on my path. With `z3` everything works fine.

/cc @gridaphobe we should definitely add a `checkExecutableOnPath` test for the SMT solver...

**Comment by formrre** [comment]
Ah alright; my bad.

**Comment by ranjitjhala** [comment]
No worries. Our bad for not having a proper error message for this case!

**Comment by gridaphobe** [comment]
Good point! Fixed in 5d4bdb9.
